proposals/ipfs-graphsync.md

```diff
@@ -0,0 +1,95 @@
+and any other information relevant for prioritizing this project against others.
+It does not need to describe the work in much detail. Most technical design and planning would take place after a proposal is adopted.
+Good project scope aims for ~3-5 engineers for 1-3 months (though feel free to suggest larger-scoped projects anyway). 
+Projects do not include regular day-to-day maintenance and improvement work, e.g. on testing, tooling, validation, code clarity, refactors for future capability, etc.
+-->
+<!--
+For ease of discussion in PRs, consider breaking lines after every sentence or long phrase.
+-->
+
+## Purpose &amp; impact 
+#### Background &amp; intent
+
+Currently, ipfs nodes speak [bitswap](https://docs.ipfs.io/concepts/bitswap/) for exchanging blocks of data. bitswap works at a block layer where you ask for a CID, and get back the block that hashes to that cid.
+
+[graphsync](https://github.com/ipfs/go-graphsync) is a protocol for synchronizing graphs of data across peers. It allows a host to make a single request to a remote peer for all of the results of traversing an IPLD selector on the remote peer's local IPLD graph. It is what filecoin uses.
+
+#### Assumptions &amp; hypotheses
+
+* There are cases where users will want to retrieve a piece of structured data they can efficiently describe with a selector, but which would be expensive to retrieve with bitswap.
+* There are cases where it would be valuable for an IPFS node and a filecoin node to exchange data.
+
+#### User workflow example
+
+* A user would ask for data using a selector in e.g. the [go-fetcher](https://github.com/ipfs/go-fetcher) interface within IPFS (from a plugin)
+* An API is built such that dag exploration of ipld nodes from a client built on IPFS is translated into selectors
+* a client builds a car of data to store, creates a filecoin deal for it, and the miner fetches the data from an IPFS client with graphsync
+
+#### Impact
+
+High. we invision cases where we want IPFS and filecoin nodes to be able to communicate
+
+#### Leverage
+
+High. we have other projects that are enabled by IPFS and filecoin nodes being able to communicate
+
+#### Confidence
+
+Two. We haven't heard direct requests for this flow, but it's our best guess at what's practical.
+
+
+## Project definition
+#### Brief plan of attack
+
+[go-merkledag](https://github.com/ipfs/go-merkledag) currently takes a [blockservice](https://github.com/ipfs/go-merkledag/blob/bf51443272bb98cff071eb44ed9ce6c940e82f1f/merkledag.go#L32).  It should also be able to be created from a graphsync provider, or both. We can initially enable graphsync as an opt-in option, and need to design a transition (e.g. do we have merkledag make requests to both graphsync and to blockservice when present? do we make this code somewhat peer-aware, so we know based on provider which one to use?)
```

Comment:
TLDR: In trying to make GraphSync usable in go-ipfs having an interface that handles both bitswap and graphsync backed DAGServices is the easy part. Content discovery, spreading requests across peers, prioritizing peers + protocols, etc. are the hard parts.

IIUC the main areas GraphSync is lacking in order to be a drop in replacement for Bitswap are:
- We have no code (or AFAIK proposal yet) for distributing data across multiple peers using GraphSync
- Graphsync is not on its own suitable for peer discovery in the way Bitswap is (e.g. it may be less efficient due to being unable to ask for HAVEs)

From what I see this limits our options, and some remaining ones are:
- Make GraphSync able to be a drop-in replacement:
  - Extract sessions code from Bitswap/create a way for sessions to be managed across Bitswap and GraphSync
  - Make sessions smart enough to split queries across peers with GraphSync, or something simpler like trying to use GraphSync only if we fail at using Bitswap for long enough
  -  Use Bitswap (or some other protocol, potentially including GraphSync) for discovery of content from peers we're already connected to
- Have options like `ipfs cat <cid> [--from=<multiaddr>]` that will use GraphSync for direct point-to-point transfers